File: static/js/date_utils.js
/**
 * Date utilities for consistent client-side date handling
 * Addresses timezone issues with JavaScript Date constructor
 */

/**
 * Parse a YYYY-MM-DD string to a Date object without timezone issues
 * @param {string} dateStr - Date string in YYYY-MM-DD format
 * @returns {Date} Date object representing the correct local date
 */
function parseDateSafe(dateStr) {
    if (!dateStr) return null;
    
    // Split the date string and create Date object safely
    const [year, month, day] = dateStr.split('-').map(Number);
    
    // Use Date constructor with individual components
    // Note: month is 0-indexed in JavaScript
    return new Date(year, month - 1, day);
}

/**
 * Format a date string for display with day of week
 * @param {string} dateStr - Date string in YYYY-MM-DD format
 * @returns {string} Formatted date like "Monday, 5/26/25"
 */
function formatDateForDisplay(dateStr) {
    if (!dateStr) return '';
    
    try {
        const date = parseDateSafe(dateStr);
        if (!date || isNaN(date.getTime())) return dateStr;
        
        const weekday = date.toLocaleDateString('en-US', { weekday: 'long' });
        const month = date.getMonth() + 1; // Convert back to 1-indexed
        const day = date.getDate();
        const year = date.getFullYear().toString().slice(-2);
        
        return `${weekday} ${month}/${day}/${year}`;
    } catch (error) {
        console.error('Error formatting date:', error);
        return dateStr;
    }
}

/**
 * Compare two date strings for equality
 * @param {string} date1 - First date string (YYYY-MM-DD)
 * @param {string} date2 - Second date string (YYYY-MM-DD)
 * @returns {boolean} True if dates are the same
 */
function isSameDate(date1, date2) {
    if (!date1 || !date2) return false;
    
    // Normalize both dates to YYYY-MM-DD format
    const normalized1 = normalizeDate(date1);
    const normalized2 = normalizeDate(date2);
    
    return normalized1 === normalized2;
}

/**
 * Normalize various date formats to YYYY-MM-DD
 * @param {string} dateStr - Date string in various formats
 * @returns {string} Date string in YYYY-MM-DD format
 */
function normalizeDate(dateStr) {
    if (!dateStr) return null;
    
    // If already in YYYY-MM-DD format, return as-is
    if (/^\d{4}-\d{2}-\d{2}$/.test(dateStr)) {
        return dateStr;
    }
    
    // Handle MM/DD/YYYY or MM/DD/YY formats
    if (dateStr.includes('/')) {
        const parts = dateStr.split('/');
        if (parts.length === 3) {
            let [month, day, year] = parts;
            
            // Convert 2-digit year to 4-digit
            if (year.length === 2) {
                year = '20' + year;
            }
            
            // Pad month and day with leading zeros
            month = month.padStart(2, '0');
            day = day.padStart(2, '0');
            
            return `${year}-${month}-${day}`;
        }
    }
    
    // If we can't parse it, return as-is
    return dateStr;
}

/**
 * Get today's date in YYYY-MM-DD format
 * @returns {string} Today's date
 */
function getTodayString() {
    const today = new Date();
    const year = today.getFullYear();
    const month = (today.getMonth() + 1).toString().padStart(2, '0');
    const day = today.getDate().toString().padStart(2, '0');
    return `${year}-${month}-${day}`;
}

/**
 * Check if a date is in the past
 * @param {string} dateStr - Date string to check
 * @returns {boolean} True if date is before today
 */
function isDateInPast(dateStr) {
    if (!dateStr) return false;
    
    const today = getTodayString();
    return normalizeDate(dateStr) < today;
}

// Export functions for use in other scripts
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        parseDateSafe,
        formatDateForDisplay,
        isSameDate,
        normalizeDate,
        getTodayString,
        isDateInPast
    };
} 

File: utils/date_utils.py
"""
Date utilities for consistent timezone-aware date handling
Addresses the timezone issues with Railway PostgreSQL
"""

from datetime import datetime, date, timezone
import pytz

# Application timezone - all dates should be interpreted in this timezone
APP_TIMEZONE = pytz.timezone('America/Chicago')

def date_to_db_timestamp(date_obj):
    """
    Convert a date object to a timezone-aware timestamp for database storage.
    After TIMESTAMPTZ migration, stores at midnight UTC for consistency.
    
    Args:
        date_obj: datetime.date, datetime.datetime, or string in various formats (YYYY-MM-DD, MM/DD/YYYY, etc.)
    
    Returns:
        datetime: Timezone-aware datetime at midnight UTC
    """
    if isinstance(date_obj, str):
        # First normalize the date string to YYYY-MM-DD format
        normalized_date_str = normalize_date_string(date_obj)
        date_obj = datetime.strptime(normalized_date_str, '%Y-%m-%d').date()
    elif isinstance(date_obj, datetime):
        date_obj = date_obj.date()
    
    # Create midnight timestamp in UTC
    midnight_dt = datetime.combine(date_obj, datetime.min.time())
    return midnight_dt.replace(tzinfo=timezone.utc)

def db_timestamp_to_date(timestamp_obj):
    """
    Convert a database timestamp back to a date object.
    
    Args:
        timestamp_obj: timezone-aware datetime from database
    
    Returns:
        date: Date object in application timezone
    """
    if timestamp_obj is None:
        return None
    
    # Convert to application timezone and extract date
    local_dt = timestamp_obj.astimezone(APP_TIMEZONE)
    return local_dt.date()

def normalize_date_string(date_str):
    """
    Normalize various date string formats to YYYY-MM-DD.
    
    Args:
        date_str: Date string in various formats
    
    Returns:
        str: Date string in YYYY-MM-DD format
    """
    if not date_str:
        return None
    
    # Try different formats
    formats = [
        '%Y-%m-%d',     # 2025-05-26
        '%m/%d/%Y',     # 05/26/2025
        '%m/%d/%y',     # 05/26/25
        '%Y/%m/%d',     # 2025/05/26
    ]
    
    for fmt in formats:
        try:
            dt = datetime.strptime(date_str, fmt)
            return dt.strftime('%Y-%m-%d')
        except ValueError:
            continue
    
    raise ValueError(f"Unable to parse date string: {date_str}")

def format_date_for_display(date_obj):
    """
    Format a date for display with day of week.
    
    Args:
        date_obj: date object or YYYY-MM-DD string
    
    Returns:
        str: Formatted date like "Monday 5/26/25"
    """
    if isinstance(date_obj, str):
        date_obj = datetime.strptime(date_obj, '%Y-%m-%d').date()
    
    day_of_week = date_obj.strftime('%A')
    date_str = date_obj.strftime('%-m/%-d/%y')  # Remove leading zeros
    return f"{day_of_week} {date_str}"

def is_same_date(date1, date2):
    """
    Compare two dates for equality, handling various input types.
    
    Args:
        date1, date2: date objects, datetime objects, or YYYY-MM-DD strings
    
    Returns:
        bool: True if dates represent the same day
    """
    def normalize_to_date(d):
        if isinstance(d, str):
            return datetime.strptime(normalize_date_string(d), '%Y-%m-%d').date()
        elif isinstance(d, datetime):
            return d.date()
        return d
    
    return normalize_to_date(date1) == normalize_to_date(date2)

# Database query helpers
def build_date_query(table_alias="", date_column="match_date"):
    """
    Build SQL for date comparison with TIMESTAMPTZ columns.
    
    After migration to TIMESTAMPTZ with midnight UTC storage, we can use simple DATE() extraction.
    
    Returns:
        str: SQL fragment like "DATE(match_date)"
    """
    prefix = f"{table_alias}." if table_alias else ""
    return f"DATE({prefix}{date_column})"

def build_date_params(date_value):
    """
    Build parameters for date queries with TIMESTAMPTZ columns.
    
    Args:
        date_value: date, datetime, or string
    
    Returns:
        datetime: Timezone-aware datetime at midnight UTC for database queries
    """
    return date_to_db_timestamp(date_value) 